Feature: Add progress bars

peterjdolan opened the pull request:
## Description

Many long-running commands produce little or no feedback in the terminal to indicate that they're progressing, and none of them provide estimates of how long the operation will run. This change introduces the `enlighten` python package, which displays progress bars akin to TQDM below the existing terminal output. This PR does not modify the import workflow, rather only modifies several built-in plugins. Modifying the import workflow to display a progress bar will be done in a follow-up PR.

See [tracking issue](https://github.com/beetbox/beets/issues/5711)

## To Do

- [x] Documentation. (If you've added a new command-line flag, for example, find the appropriate page under `docs/` to describe it.)
- [x] Changelog. (Add an entry to `docs/changelog.rst` to the bottom of one of the lists near the top of the document.)
- [x] Tests. (Very much encouraged but not strictly required.)

## Example screenshot

![Screenshot 2025-04-10 103457](https://github.com/user-attachments/assets/be6e8146-9476-4f21-9e41-683e8779c6c0)

peterjdolan committed:
Add progress bars to several commands and plugins.

Many long-running commands produce little or no feedback in the terminal to
indicate that they're progressing, and none of them provide estimates of how
long the operation will run. This change introduces the `enlighten` python
package, which displays progress bars akin to TQDM below the existing terminal
output.

To support consistent use and presentation of the progress bars, and to
allow for future modification, we introduce a method to beets.ui -
beets.ui.iprogress_bar - which can be used by Beets' core commands and all
Beets plugins. Example usage is provided in the methods' documentation
and in a new 'further reading' doc for developers.

The Enlighten library does not work as well in Windows PowerShell as it
does in a linux terminal (manually tested in Zsh), so the progress bars
are disabled in Windows environments. Resolving these issues and enabling
them in Windows is left as future work.

Progress Bars are disabled also when not in a TTY, to prevent
interference with logging when running as a web server, for example.

peterjdolan committed:
Add progress bars to several commands and plugins.

Many long-running commands produce little or no feedback in the terminal to
indicate that they're progressing, and none of them provide estimates of how
long the operation will run. This change introduces the `enlighten` python
package, which displays progress bars akin to TQDM below the existing terminal
output.

To support consistent use and presentation of the progress bars, and to
allow for future modification, we introduce a method to beets.ui -
beets.ui.iprogress_bar - which can be used by Beets' core commands and all
Beets plugins. Example usage is provided in the methods' documentation
and in a new 'further reading' doc for developers.

The Enlighten library does not work as well in Windows PowerShell as it
does in a linux terminal (manually tested in Zsh), so the progress bars
are disabled in Windows environments. Resolving these issues and enabling
them in Windows is left as future work.

Progress Bars are disabled also when not in a TTY, to prevent
interference with logging when running as a web server, for example.

peterjdolan commented:
Import operation partially implemented; still needs some work. Example screenshot above.

peterjdolan commented:
@snejus I'm not sure why the latest push is failing the automatic lint and formatting; it passes 'poetry run poe lint && poetry run poe format' on my machine just fine. Any suggestions? I've already sync'd my fork to beets master and rebased.

wisp3rwind commented:
This seems to be related to some combination of (although I'm not seeing the full picture yet)
- you regenerated the lockfile, which now chooses a newer version of ruff
- https://github.com/astral-sh/ruff/blob/6bdffc3cbff2194afdff0e46fccb03cf6147d367/CHANGELOG.md?plain=1#L987
- https://github.com/astral-sh/ruff/pull/14511
- maybe, you're still running another ruff version locally

Someone should probably open a new PR to fix all TC006 warnings, include `TC` in our ruff rule selection instead of TCH, and probably fix ruff to a major version (a `0.X.*`). Since major ruff updates might contain breaking rule/style changes, I'd guess that we'd only want to bump it manually.

jackwilsdon commented:
It looks like this needs to replicate the check `enlighten.get_manager` does - disabling the manager if the output is not a TTY.

You might be best actually just using `enlighten.get_manager` and passing enabled to that, as it will only enable the manager if the provided enable value was true and the stream is a TTY.

peterjdolan commented:
This is now ready for review. I'm not certain how to request a review, though. Anybody have pointers?

snejus reviewed (changes requested):
Added a fair bit of comments

peterjdolan committed:
Add progress bar to 'beet mv', and handle missing file exceptions without crashing.

peterjdolan committed:
Add progress bar to 'beet mv', and handle missing file exceptions without crashing.

peterjdolan committed:
Add progress bars to 'beet update', 'beet rm', 'beet mv', and the modify command.

peterjdolan committed:
Add progress bars to 'beet update', 'beet rm', 'beet mv', and the modify command.

peterjdolan commented on beets/ui/__init__.py:
```diff
@@ -1325,6 +1344,80 @@ def add_all_common_options(self):
         self.add_format_option()
 
 
+T = TypeVar("T", covariant=True)
```
Done

peterjdolan commented on beets/ui/__init__.py:
````diff
@@ -1325,6 +1344,80 @@ def add_all_common_options(self):
+    Example usage:
+    ```
+    for album in ui.iprogress_bar(
+        lib.albums(), desc="Updating albums", unit="albums"):
````
Done

peterjdolan commented on beets/ui/__init__.py:
```diff
@@ -1325,6 +1344,80 @@ def add_all_common_options(self):
+    if sequence is None:
+        log.error("sequence must not be None")
+        return
+
```
Done

peterjdolan committed:
Code review feedback

- Add lib.items_with_progress, lib.albums_with_progress
- Code formatting and style
- Update docs

peterjdolan committed:
Code review feedback

- Add lib.items_with_progress, lib.albums_with_progress
- Code formatting and style
- Update docs

peterjdolan committed:
Revise documentation formatting

peterjdolan committed:
Revise documentation formatting

peterjdolan commented on beetsplug/missing.py:
```diff
@@ -157,13 +156,13 @@ def _missing_tracks(self, lib, query):
-
-            else:
+        else:
+            for album in albums:
```
Ah, my mistake. I had misread the original logic and thought there was a bug, but I understand it now. Reverted.

peterjdolan commented on beetsplug/parentwork.py:
```diff
@@ -88,12 +88,18 @@ def func(lib, opts, args):
-                    if write:
-                        item.try_write()
+
+                if write:
```
My mistake, thank you.

peterjdolan commented on beetsplug/permissions.py:
```diff
@@ -31,10 +31,13 @@ def check_permissions(path, permission):
     `check_permissions`.
+
+    Returns:
+        True if the permissions are as expected, False otherwise.
```
De-redundantized :)

peterjdolan committed:
Review feedback

peterjdolan committed:
Review feedback

peterjdolan commented on pyproject.toml:
```diff
@@ -81,6 +81,7 @@ pydata-sphinx-theme = { version = "*", optional = true }
 sphinx = { version = "*", optional = true }
 sphinx-design = { version = "^0.6.1", optional = true }
 sphinx-copybutton = { version = "^0.5.2", optional = true }
+enlighten = "^1.14.1"
```
Done

peterjdolan committed:
Review feedback

peterjdolan committed:
Review feedback

peterjdolan commented on poetry.lock:
```diff
@@ -1,11 +1,13 @@
-# This file is automatically @generated by Poetry 1.8.5 and should not be changed by hand.
+# This file is automatically @generated by Poetry 2.2.1 and should not be changed by hand.
```
Done, thank you.

peterjdolan committed:
Fix poetry, and lint

peterjdolan committed:
Fix poetry, and lint

peterjdolan committed:
Formatting

peterjdolan committed:
Formatting

peterjdolan committed:
Plural units -> singular units

peterjdolan committed:
Plural units -> singular units

peterjdolan committed:
Fix units

peterjdolan committed:
Fix units

peterjdolan committed:
Fix tests

peterjdolan committed:
Fix tests

peterjdolan reviewed (commented):
I've addressed your comments to the best of my ability, and left unresolved the comments that I had questions on. Thanks for your review!

peterjdolan committed:
Formatting

peterjdolan committed:
Formatting

peterjdolan committed:
Fix small counter error

peterjdolan committed:
Fix small counter error

peterjdolan committed:
Fix import

peterjdolan committed:
Fix import

peterjdolan committed:
Merge branch 'tqdm' of https://github.com/peterjdolan/beets into tqdm

peterjdolan commented:
@snejus it seems that 'poe format-docs' reformats virtually all of the docs; should I do so, or would you like another PR that does so that I can submit first?

peterjdolan committed:
Format progressbars.rst

peterjdolan committed:
Revert unintended change to format-docs command

peterjdolan committed:
Check if revision is not None, as it will be 0 for items not net written to the database.

masroof-maindak commented:
Many thanks to everyone involved.

Could we please get an expected date for when this will be merged?

~If it's a while in the future, does anyone know an easy way to install this fork using `uv`?~ Edit: the `--from` flag.
